### ISSUE TITLE
Some questions about the paper

### ISSUE THREAD

**Issue by kwang2049** [princeton-nlp/SimCSE#1]
The paper is really interesting. Appreciate your hard work.

I have two questions:
1. Really good insight about using Alignment and Uniformity (A&U) to compare and analyze different approaches. However, given that different models might have different similarity ranges, so is the absolute A&U value of different models really comparable to each other(e.g. in Figure 3)?
2. What is the training data for SimCSE in each experiment? I did not find the corresponding texts (maybe I missed them).

**Comment by danqi** [comment]
1. All the vectors are normalized (the sentence before Equation 2), which means all the similarity scores range between -1 and 1.
2. For unsupervised SimCSE, we use 1 million sentences randomly drawn from English Wikipedia. For supervised SimCSE, we use 314k entailment pairs and 314k contradiction pairs from SNLI + MNLI. They are mentioned in Sections 3 & 4. We will emphasize them again in the experiment section.

**Comment by kwang2049** [comment]
Thanks for your reply. Sorry I didn't manage myself to keep track of this discussion.

About the first question, yes, the cosine similarity confines the score range within [-1, 1]. What I was really unclear is this case: 

```
Gold labels for 5 examples: 1, 2, 3, 4, 5
Prediction of model 1: 0.1, 0.2, 0.3, 0.4, 0.5
Prediction of model 2: 0.6, 0.7, 0.8, 0.9, 1.0
```

Both models will get a perfect score of 100% Spearman correlation. It also seems to make sense since only the (consistent) ranks matter and one might not care about how a model defines what absolute scores represents the most similar/dissimilar.

But this case would be wrongly captured by using Alignment: model2 > model1.

So do you think I am right? Or any misunderstanding of mine?

**Comment by gaotianyu1350** [comment]
Hi,

You made a good point. In fact, the metric alignment (or uniformity) cannot be used alone. Another example is that, vanilla BERT or RoBERTa have almost perfect alignment (because the embeddings are very anisotropic), but they perform really bad.

So alignment and uniformity should be used in a combined way: if both alignment and uniformity are good, then the model is likely to perform well. Either bad uniformity or bad alignment will lead to performance collapse. 

Another problem might be that even though two models are equally good, model 2 could have (absolute value) better alignment and uniformity. This can happen, but since all our models in experiments use the same pre-trained models as initialization, the absolute value distributions of cosine similarity don't differ that much. Also, for Figure 2, where we portray how models' alignment/uniformity evolve with training, this is not a big problem since we can see a clear trend that models are achieving better balance between alignment/uniformity (regardless of what the absolute values are).

**Comment by kwang2049** [comment]
> Hi,
> 
> You made a good point. In fact, the metric alignment (or uniformity) cannot be used alone. Another example is that, vanilla BERT or RoBERTa have almost perfect alignment (because the embeddings are very anisotropic), but they perform really bad.
> 
> So alignment and uniformity should be used in a combined way: if both alignment and uniformity are good, then the model is likely to perform well. Either bad uniformity or bad alignment will lead to performance collapse.
> 
> Another problem might be that even though two models are equally good, model 2 could have (absolute value) better alignment and uniformity. This can happen, but since all our models in experiments use the same pre-trained models as initialization, the absolute value distributions of cosine similarity don't differ that much. Also, for Figure 2, where we portray how models' alignment/uniformity evolve with training, this is not a big problem since we can see a clear trend that models are achieving better balance between alignment/uniformity (regardless of what the absolute values are).

Totally agree with you on Figure2. I really like that Figure and also think A&U for dynamics analysis is a perfect use case. 

About the range issue, we can think a simpler example:
Model 1 generates whatever similarity scores;
model 2 generates exactly the model 1's similarity + 0.5.
This time, model 1 < model 2 for alignment and model 2 > model 1 for uniformity according to Equation (2) and (3). However, we have no clue about which one is better by observing just A&U (although we can cheat by observing Spearman to see they are actually equally good).

In fact, one might consider applying some LayerNorm like normalization to prevent this from happening?

**Comment by gaotianyu1350** [comment]
This indeed could happen. The uniformity/alignment metrics are mainly used for analysis, and we still rely on Spearman's correlation on the test set for measuring the sentence embedding quality. Thus even though there might be ways to fix the potential problems, I feel it is ok to leave uniformity/alignment as what they are now.
